### PR TITLE
nimble/transport: Refactor HCI transport

### PR DESCRIPTION
**Pull request by andrzej-kaczmarek**
This is a huge refactor of HCI transport ot make it - hopefully - easier to maintain and configure.

There's now common code that manages cmd/evt/ACL pools, no need to do this is each transport separately.
HS and LL no longer use callbacks to register to transport, instead there are dedicated APIs which they use and link directly and thus there's no need for separate RAM transport since both host and controller and host will call directly each other.
Each transport can implement those APIs and thus act as host or controller side, this makes possible to flexibly select transport on each side and thus removes need for dedicated bridge app - blehci can to the same now.
Settings and automatic package dependencies were also simplified - app can only include nimble/host and all dependencies and configuration will be applied automatically assuming BSP is properly configured.

See nimble/doc/transport.md for more information.

### CODE REVIEW

File: nimble/transport/common/hci_h4/src/hci_h4.c [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+#include <assert.h>
+#include <stdint.h>
+#include <string.h>
+#include <syscfg/syscfg.h>
+#include <os/os_mbuf.h>
+#include <nimble/hci_common.h>
+#include <nimble/transport.h>
+#include <nimble/transport/hci_h4.h>
+
+#define MIN(_a, _b) ((_a) < (_b) ? (_a) : (_b))
```
Comment (sjanc) [inline nimble/transport/common/hci_h4/src/hci_h4.c:29]: we have min() in os/os.h

File: nimble/transport/common/hci_h4/src/hci_h4.c [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+const struct hci_h4_allocators hci_h4_allocs_from_hs = {
+    .cmd = ble_transport_alloc_cmd,
+    .acl = ble_transport_alloc_acl_from_hs,
+};
+
+struct hci_h4_input_buffer {
+    const uint8_t *buf;
+    uint16_t len;
+};
+int andk;
```
Comment (sjanc) [inline nimble/transport/common/hci_h4/src/hci_h4.c:49]: :-)

File: nimble/transport/common/hci_h4/src/hci_h4.c [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+            h4sm->state = HCI_H4_SM_COMPLETED;
+        /* no break */
+        case HCI_H4_SM_COMPLETED:
+            hci_h4_sm_completed(h4sm);
+            h4sm->state = HCI_H4_SM_W4_PKT_TYPE;
+            break;
+        default:
+            assert(0);
+            /* consume all remaining data */
+            hci_h4_ib_consume(&ib, ib.len);
```
Comment (sjanc) [inline nimble/transport/common/hci_h4/src/hci_h4.c:279]: wouldn't this already assert?

File: nimble/transport/src/transport.c [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+#include <stdint.h>
+#include <syscfg/syscfg.h>
+#include <sysinit/sysinit.h>
+#include <os/os_mbuf.h>
+#include <os/os_mempool.h>
+#include <nimble/ble.h>
+#include <nimble/hci_common.h>
+#include <nimble/transport.h>
+
+os_mempool_put_fn *transport_put_acl_from_ll_cb;
```
Comment (sjanc) [inline nimble/transport/src/transport.c:29]: can this be static?

File: nimble/controller/src/ble_ll_conn_hci.c [hunk 5]
```diff
@@ -76,7 +75,7 @@ ble_ll_init_alloc_conn_comp_ev(void)
     rc = 0;
     evbuf = g_ble_ll_conn_comp_ev;
     if (evbuf == NULL) {
-        evbuf = ble_hci_trans_buf_alloc(BLE_HCI_TRANS_BUF_EVT_HI);
+        evbuf =ble_transport_alloc_evt(0);
```
Comment (sjanc) [inline nimble/controller/src/ble_ll_conn_hci.c:78]: I liked how original defines for hi/low made code easy to read,   maybe add BLE_HCI_TRANS_EVT_DISCARDABLE or similar?

File: nimble/host/syscfg.yml [hunk 6]
```diff
@@ -497,3 +497,6 @@ syscfg.logs:
 
 syscfg.vals.BLE_MESH:
     BLE_SM_SC: 1
+
+syscfg.restrictions:
+    - BLE_TRANSPORT_HS == "native"
```
Comment (sjanc) [inline nimble/host/syscfg.yml:464]: hmm I'm not sure if I get this restriction

File: nimble/transport/emspi/src/ble_hci_emspi.c [hunk 7]
```diff
@@ -955,3 +684,15 @@ ble_hci_emspi_init(void)
...
+ble_transport_to_ll_cmd(void *buf)
+{
+    return ble_hci_emspi_cmdevt_tx(buf, BLE_HCI_EMSPI_PKT_CMD);
+}
+
+int
+ble_transport_to_ll_acl(struct os_mbuf *om)
+{
+    return ble_hci_emspi_acl_tx(om);
+}
```
Comment (sjanc) [inline nimble/transport/emspi/src/ble_hci_emspi.c:698]: nitpick

File: babblesim/edtt/hci_transport/syscfg.yml [hunk 8]
```diff
@@ -31,19 +31,10 @@ syscfg.defs:
...
-    BLE_ACL_BUF_SIZE: 255
-    BLE_HCI_EVT_BUF_SIZE: 257
-    BLE_HCI_EVT_HI_BUF_COUNT: 32
-    BLE_HCI_EVT_LO_BUF_COUNT: 32
+    BLE_TRANSPORT_ACL_COUNT: 32
+    BLE_TRANSPORT_EVT_COUNT: 64
+
+syscfg.restrictions:
+    - BLE_TRANSPORT_HS == "custom"
+    - BLE_TRANSPORT_LL == "native"
```
Comment (sjanc) [inline babblesim/edtt/hci_transport/syscfg.yml:40]: nitpick

File: nimble/host/test/src/ble_hs_test_util.c [hunk 9]
```diff
@@ -2046,3 +2041,9 @@ ble_hs_test_util_init(void)
     /* Clear random address. */
     ble_hs_id_rnd_reset();
 }
+
+void
+ble_transport_ll_init(void)
+{
+    /* nothing here */
+}
```
Comment (sjanc) [inline nimble/host/test/src/ble_hs_test_util.c:2049]: nitpick

**Review by sjanc — COMMENTED** [review]
nice :)

File: nimble/controller/src/ble_ll_conn_hci.c (hunk 5 above)
Comment (andrzej-kaczmarek) [reply nimble/controller/src/ble_ll_conn_hci.c:78]: the parameter name is `discardable` so it's pretty clear what `0` and `1` mean here

File: nimble/host/syscfg.yml (hunk 6 above)
Comment (andrzej-kaczmarek) [reply nimble/host/syscfg.yml:464]: this is to make sure transport on hs side is not overridden to something else when an actual host is included in build - such conbination would not make sense

File: nimble/transport/src/transport.c (hunk 4 above)
Comment (andrzej-kaczmarek) [reply nimble/transport/src/transport.c:29]: I'll investigate and get back to you

File: nimble/transport/src/transport.c [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+static struct os_mbuf_pool mpool_acl;
+
+void *
+ble_transport_alloc_cmd(void)
+{
+    return os_memblock_get(&pool_cmd);
+}
+
+void *
+ble_transport_alloc_evt(int discarcable)
```
Comment (kasjer) [inline nimble/transport/src/transport.c:83]: typo

**Review by kasjer — COMMENTED** [review]
maybe targets for cmac and net core of nrf5340 also should be updated

**Review by sjanc — APPROVED** [review]

**Pull request merged**